[Speculation] Add a basic Pass that reads from a JSON file

asleix opened the pull request:
Speculative execution involves speculating on the output of a long latency operation or chain of operations, allowing operations that depend on it to begin execution early, without impacting correctness.

The pass HandshakeSpeculation is implemented, which places the Speculation Operations according to the positions specified in an input JSON file. A helper class called SpeculationPlacements holds the placement information and implements the JSON parser.

This pull-request (PR) includes the basic structure and some dummy connections, and will be followed by 1-2 PRs implementing more features. The idea is the following.

**PR 1 (this one).** The pass will place the units according to an input JSON file, and use simple connections.
**PR 2**. Use the correct control path for connecting Commits and SaveCommits. Still read position from JSON.
**PR 3**. If no JSON is provided, automatically choose the positions for placing the units.

The work for PR2 and PR3 is mostly done and being tested. The split into smaller merges should help reviewing. I will wait for PR1 to be merged to create PR2 and PR3. 

The pass can be run with `dynamatic-opt input.mlir --handshake-speculation="unit-positions=positions.json"`An input JSON should have the following format. This example can be used in the handshake IR of `integration-test/fir`.
```
{
  "speculator": {"srcOp": "cmpi0", "dstOp":"fork5"},
  "saves": [
    {"srcOp": "extsi6", "dstOp":"mc_load0"},
    {"srcOp": "extsi7", "dstOp":"mc_load1"}
  ],
  "commits": [
    {"srcOp": "trunci0", "dstOp":"cond_br0"}
  ],
  "saveCommits": [
    {"srcOp": "mux1", "dstOp":"buffer10"}
  ]
}
```

For PR2, example JSON files will be explicitly added because they will be needed for unit tests of the pass. 

**Important:** A newer version of CIRCT is required for this to work https://github.com/EPFL-LAP/circt/pull/1.

murphe67 commented on lib/Transforms/Speculation/SpeculationPlacement.cpp:
```diff
@@ -0,0 +1,252 @@
+  while (not llvm::is_contained(dstOp->getOperands(), srcOpResult)) {
+    result_idx++;
+    if (result_idx >= srcOp->getNumResults()) {
+      return failure(); // srcOp and dstOp are not connected
```
This is an error in an end user's input file, so you should probably output the failure reason.

murphe67 commented on lib/Transforms/Speculation/SpeculationPlacement.cpp:
```diff
@@ -0,0 +1,252 @@
+  while (not llvm::is_contained(dstOp->getOperands(), srcOpResult)) {
+    result_idx++;
+    if (result_idx >= srcOp->getNumResults()) {
+      return failure(); // srcOp and dstOp are not connected
```
This is an error in an end user's input file, so you should probably output the failure reason.

asleix replied:
Solved in the new commit.

murphe67 commented on lib/Transforms/Speculation/HandshakeSpeculation.cpp:
```diff
@@ -0,0 +1,101 @@
+}
+
+std::unique_ptr<dynamatic::DynamaticPass>
+dynamatic::speculation::createHandshakeSpeculation(std::string unitPositions) {
```
I would rename unitPositions to jsonPath to be consistant throughout the whole project, and also to be clearer.

asleix replied:
I agree that the name does not reveal that the string is a path to a json file. However, the name comes from the pass argument in Passes.td
`Option<"unitPositions", "unit-positions", "std::string", "",
    "Path to JSON-formatted file containing the positions for speculative "
    "operations.">]`. So the replacement name for "unitPositions" should be consistent with the pass argument "unit-positions", or replace the two altogether. What is your suggestion?

murphe67 replied:
When you pass it to readFromJSON it becomes jsonPath. I think it should be jsonPath from start to finish, including in the pass argument, and at least shouldn't switch names.

murphe67 commented on lib/Transforms/Speculation/HandshakeSpeculation.cpp:
```diff
@@ -0,0 +1,101 @@
+  if (failed(placeSpeculator()))
+    return signalPassFailure();
+
+  // Place Save operations
```
I would replace this->specOp->getResult(n) with named accessor functions, its easy to type the wrong number somewhere and get a bug, and its easier if you ever change the numbers.

asleix replied:
I agree that named accessor functions could help a lot to improve the clarity. However, operations do not have that by default. Explicit accessor functions should be implemented in CIRCT first. I suggest to postpone that for a future improvement.

murphe67 replied:
Positional at the IR level is standard and almost always fine, it's only based on the complexity of the speculators outputs, (and routing them)  that I would wrap them with named outputs in the pass itself.
Just to avoid having to go look in another file, while trying to remember the specifics of the 3 auxiliary control signals.

asleix replied:
At first I thought that get<...> was a templated function. It took me a bit to see what you meant.
Done: now using the named accessor functions.

murphe67 reviewed (commented):
Looks great, thanks for updating everything we talked about! :). Added some small recommendations.

RamirezLucas commented on lib/Transforms/Speculation/SpeculationPlacement.cpp:
```diff
@@ -0,0 +1,255 @@
+  }
+
+  // Add Save Operations position
+  for (PlacementNames &p : specNameMap["saves"]) {
```
Use a lambda to avoid duplicating basically the same piece of code 4 times.

asleix replied:
Done.

RamirezLucas commented on lib/Transforms/Speculation/SpeculationPlacement.cpp:
```diff
@@ -0,0 +1,255 @@
+  // Add Save Operations position
+  for (PlacementNames &p : specNameMap["saves"]) {
+    Operation *srcOp = nameAnalysis.getOp(p.srcOpName);
+    Operation *dstOp = nameAnalysis.getOp(p.dstOpName);
```
There needs to be some error checking there, in case any of the operations denoted by the names in the JSON do not exist. Something like:
```cpp
if (!srcOp || !dstOp)
  return failure();
```

asleix replied:
Done.

RamirezLucas commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+  StringRef dstOpName;
+};
+
+struct OpPlacement {
```
With the JSON format change I propose, this struct can probably just be replaced by an `mlir::OpOperand&` (a reference to an operation's operand i.e., a value use).

asleix replied:
For convenience, now I am using a struct that is storing the `opName` as a string and `opIdx` as an unsigned integer.

RamirezLucas commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+class SpeculationPlacements {
+private:
+  OpPlacement speculator;
+  PlacementList saves;
```
These do not seem to be used as sets at all (e.g., no `contains` query as far as I see). If not, a vector is simpler and better (`SmallVector<OpOperand&>` with the `OpPlacement` change described above).

asleix replied:
The sets are being in code that is yet to be pushed, in methods `containsSave`, `containsCommit`, `eraseSave` and `eraseCommit`. A vector would make the implementation of these methods a bit longer. I suggest to keep using the set, what do you think?

RamirezLucas replied:
Sounds good, thanks for clarifying. You can keep them as is then.

asleix commented on include/dynamatic/Transforms/Speculation/HandshakeSpeculation.h:
```diff
@@ -0,0 +1,63 @@
+  LogicalResult placeSpeculator();
+
+  // Place the operation specified in T with the control signal ctrlSignal
+  template <typename T>
```
The templated function `placeUnits` and the function `placeSpeculator` are now private.

asleix commented on include/dynamatic/Transforms/Speculation/HandshakeSpeculation.h:
```diff
@@ -0,0 +1,63 @@
+#include "dynamatic/Transforms/Passes.h.inc"
+
+/// Public pass driver for the speculation placement pass.
+struct HandshakeSpeculationPass
```
Done: the pass driver is no longer public.

asleix commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+  PlacementList saveCommits;
+
+public:
+  // Empty constructor
```
Done.

asleix commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+  };
+};
+
+typedef std::unordered_set<OpPlacement, OpPlacement::Hash> PlacementList;
```
Done. 
My intellisense on VSCode is not reporting this. I tried changing configurations but didn't manage to get it to work. I apologize if I push code with this kind of reports.

RamirezLucas replied:
I use this VSCode extension, in case you want to try it llvm-vs-code-extensions.vscode-clangd

asleix replied:
Thanks! I'll try to set it up.

asleix commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+
+public:
+  // Empty constructor
+  SpeculationPlacements(){};
```
Done.

asleix commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+  SpeculationPlacements(){};
+
+  // Initializer with source and destination operations for the Speculator
+  SpeculationPlacements(Value srcOpResult, Operation *dstOp) {
```
Done.

asleix commented on include/dynamatic/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,108 @@
+  OpPlacement getSpeculatorPlacement();
+
+  // Get a vector of the existing Save operation placements
+  template <typename T>
```
In this case, a general templated function is never defined, but rather only three specializations, and the three should be public. I think that in this case, it would be best to leave the specializations in the .cpp. Otherwise, how would you define these in the .h file?

RamirezLucas replied:
NB: After looking online I think it would in fact work as-is compile/link-wise, but my last point stands nonetheless.

RamirezLucas replied:
NB 2: Just realized that `placeUnits<T>` requires that `getPlacements<T>` is templated too to work elegantly... I understand now why you did it this way. Still not super happy with getPlacements<T> being templated due to the reason above but feel free to ignore all of this then, sorry for the endless rant.

asleix replied:
I understand your point, thanks for the detailed reasoning. At first, I had created 3 non-templated getter methods. However, as you noticed, when writing `placeUnits<T>` I thought about the possibility of creating a templated getter. I will leave it like this for now, we can rethink about it in the following PRs.

asleix commented on include/dynamatic/Transforms/Speculation/HandshakeSpeculation.h:
```diff
@@ -0,0 +1,63 @@
+namespace speculation {
+
+std::unique_ptr<dynamatic::DynamaticPass>
+createHandshakeSpeculation(std::string unitPositions = "");
```
Done. I tried usign StringRef but then got compilation errors when usign ` std::ifstream inputFile(jsonPath);`. As a result, `const std::string &` are now used.

asleix replied:
Thanks for pointing this out! Another reason for using `const std::string &` is that the BackAnnotate pass is using this same type, which makes the pass design more homogeneous with the existing ones.

RamirezLucas replied:
Yeah I am myself not very consistent on this one (especially on older code). `StringRef` is, I think, the canonical way of efficiently passing a string you don't intend to modify in the LLVM world, and avoids a verbose-ish `const`. In any case, it's fine :)

asleix commented on lib/Transforms/Speculation/SpeculationPlacement.cpp:
```diff
@@ -0,0 +1,255 @@
+  return this->saveCommits;
+}
+
+inline void parseSpeculatorPlacement(
```
Done.

asleix commented on lib/Transforms/Speculation/SpeculationPlacement.cpp:
```diff
@@ -0,0 +1,255 @@
+    std::map<StringRef, llvm::SmallVector<PlacementNames>> &specNameMap,
+    NameAnalysis &nameAnalysis) {
+  // Add Speculator operation position
+  {
```
Done.

asleix commented:
Thanks a lot for the very detailed review!

I have added three commits to address all the comments. 
- Format, C++ correctness and code arrangement.
- Changing the JSON format to the new proposed format with operand indices.
- Moving the code to `experimental`

Here is an example of the new JSON format (it's the same example I shared in the previous format).
```
{
  "speculator": {
    "operation-name": "fork5",
    "operand-idx": 0
  },
  "saves": [
    {
      "operation-name": "mc_load0",
      "operand-idx": 0
    },
    {
      "operation-name": "mc_load1",
      "operand-idx": 0
    }
  ],
  "commits": [
    {
      "operation-name": "cond_br0",
      "operand-idx": 1
    }
  ],
  "saveCommits": [
    {
      "operation-name": "buffer10",
      "operand-idx": 0
    }
  ]
}
```

I think I addressed all the comments. Please tell me if I missed anything.

RamirezLucas commented on experimental/include/experimental/Transforms/Passes.td:
```diff
@@ -29,4 +29,26 @@ def HandshakeFixArgNames : DynamaticPass<"handshake-fix-arg-names"> {
+    "Path to JSON-formatted file containing the positions for speculative "
+    "operations.">];
+}
+
```
nit: add empty line at the end of the file

asleix replied:
Done.

RamirezLucas commented on experimental/include/experimental/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -0,0 +1,109 @@
+namespace speculation {
+
+struct PlacementOperand {
+  StringRef opName;
```
Wondering whether it is safe to store only a reference to the string here memory-wise. Do you have the guarantee that the underlying memory region holding the string is still there when you end up using the struct? If the string was parsed from the JSON file, I don't think it would be. If that's true, replace this with a `std::string`.

asleix replied:
Replaced with `std::string` because I agree that the problem that you are mentioning could happen.

RamirezLucas reviewed (commented):
Thank you for addressing everything very quickly and nicely. I have left a few comments following your changes (some new, some as followups to previous ones) which should be quick to address. After that it will be good to merge :)

asleix commented:
I implemented your latest comments and updated the CIRCT branch. I also merged the main branch and checked that the pass worked. And I want to let you know that I have prepared unit tests that I will include in the second PR for this project. Thanks for your quick feedback!

RamirezLucas reviewed (approved):
LGTM, thanks for all the work!
